client/src/in_mem.rs

```diff
@@ -0,0 +1,208 @@
+				BlockchainStorage {
+					blocks: HashMap::new(),
+					hashes: HashMap::new(),
+					best_hash: HeaderHash::default(),
+					best_number: 0,
+					genesis_hash: HeaderHash::default(),
+				})
+		}
+	}
+
+	fn insert(&self, hash: HeaderHash, header: block::Header, body: Option<block::Body>, is_new_best: bool) {
+		let number = header.number;
+		let mut storage = self.storage.write();
+		storage.blocks.insert(hash, Block {
+			header: header,
+			body: body,
+		});
+		storage.hashes.insert(number, hash);
+		if is_new_best {
+			storage.best_hash = hash;
+			storage.best_number = number;
+		}
+		if number == 0 {
+			storage.genesis_hash = hash;
+		}
+	}
+}
+
+impl blockchain::Backend for Blockchain {
+	fn header(&self, id: BlockId) -> error::Result<Option<block::Header>> {
+		Ok(self.id(id).and_then(|hash| self.storage.read().blocks.get(&hash).map(|b| b.header.clone())))
+	}
+
+	fn body(&self, id: BlockId) -> error::Result<Option<block::Body>> {
+		Ok(self.id(id).and_then(|hash| self.storage.read().blocks.get(&hash).and_then(|b| b.body.clone())))
+	}
+
+	fn info(&self) -> error::Result<blockchain::Info> {
+		let storage = self.storage.read();
+		Ok(blockchain::Info {
+			best_hash: storage.best_hash,
+			best_number: storage.best_number,
+			genesis_hash: storage.genesis_hash,
+		})
+	}
+
+	fn status(&self, id: BlockId) -> error::Result<BlockStatus> {
+		match self.id(id).map_or(false, |hash| self.storage.read().blocks.contains_key(&hash)) {
+			true => Ok(BlockStatus::InChain),
+			false => Ok(BlockStatus::Unknown),
+		}
+	}
+
+	fn hash(&self, number: block::Number) -> error::Result<Option<block::HeaderHash>> {
+		Ok(self.id(BlockId::Number(number)))
+	}
+}
+
+impl backend::Transaction for Transaction {
+	type State = state_machine::backend::InMemory;
+
+	fn state(&self) -> error::Result<Self::State> {
+		Ok(self.pending_state.clone())
+	}
+
+	fn import_block(&mut self, header: block::Header, body: Option<block::Body>, is_new_best: bool) -> error::Result<()> {
+		assert!(self.pending_block.is_none(), "Only one block per transaction is allowed");
+		self.pending_block = Some(PendingBlock {
+			block: Block {
+				header: header,
+				body: body,
+			},
+			is_best: is_new_best,
+		});
+		Ok(())
+	}
+}
+
+/// In-memory backend. Keeps all states and blocks in memory. Useful for testing.
+pub struct Backend {
+	states: RwLock<HashMap<block::HeaderHash, state_machine::backend::InMemory>>,
+	blockchain: Blockchain,
+}
+
+impl Backend {
+	/// Create a new instance of in-mem backend.
+	pub fn new() -> Backend {
+		Backend {
+			states: RwLock::new(HashMap::new()),
+			blockchain: Blockchain::new(),
+		}
+	}
+}
+
+impl backend::Backend for Backend {
+	type Transaction = Transaction;
+	type Blockchain = Blockchain;
+	type State = state_machine::backend::InMemory;
+
+	fn begin_transaction(&self, block: BlockId) -> error::Result<Self::Transaction> {
+		let state = match block {
+			BlockId::Hash(h) if h.is_zero() => Self::State::default(),
```

Comment:
why the special case for `h.is_zero()`? shouldn't that just fail?

Reply:
This is a special case for writing the genesis block.